bindings/bind_py_structure_manager.cc

```diff
@@ -616,6 +618,196 @@ namespace rascal {
+            for (py::handle key_l : all_keys_l) {
+              auto key = py::cast<std::vector<int>>(key_l);
+              all_keys.insert(key);
+            }
+          } else {
+            // empty all_keys_l means that we use the keys from the managers
+            for (auto key_l : managers.get_keys(calculator)) {
+              all_keys.insert(key_l);
+            }
+          }
+
+          auto property_name{managers.get_calculator_name(calculator, true)};
+
+          const auto & property_ =
+              *managers[0]->template get_property<Prop_t>(property_name);
+          // assume inner_size is consistent for all managers
+          int inner_size{property_.get_nb_comp() / ThreeD};
+
+          math::Matrix_t features{};
+          auto n_rows{ThreeD * managers.template get_number_of_elements<Prop_t>(
+                                   property_name)};
+          size_t n_cols{all_keys.size() * inner_size};
+          features.resize(n_rows, n_cols);
+          features.setZero();
+
+          int i_row{0};
+          for (auto & manager : managers) {
+            const auto & property =
+                *manager->template get_property<Prop_t>(property_name);
+            auto n_rows_manager = property.size() * ThreeD;
+            property.fill_dense_feature_matrix_gradient(
+                features.block(i_row, 0, n_rows_manager, n_cols), all_keys);
+            i_row += n_rows_manager;
+          }
+
+          return features;
+        },
+        R"(Get the dense feature matrix associated with the calculator and
+        the collection of structures (managers) using the list of keys
+        provided. Only applicable when Calculator uses BlockSparseProperty.)");
+    manager_collection.def(
+        "get_atoms_for_predictions",
+        [](ManagerCollection_t & managers) {
+          if (managers.size() == 0) {
+            throw std::runtime_error(
+                R"(There are no structure to get features from)");
+          }
+          auto n_atoms{0};
+          for (auto & manager : managers) {
+            n_atoms += manager->size();
+          }
+
+          Eigen::Matrix<int, Eigen::Dynamic, 3> ij_mat{};
+          ij_mat.resize(n_atoms, 3);
+          ij_mat.setZero();
+          int i_center{0}, i_frame{0};
+          for (auto & manager : managers) {
+            for (auto center : manager) {
+              ij_mat(i_center, 0) = i_frame;
+              ij_mat(i_center, 1) = i_center;
+              ij_mat(i_center, 2) = center.get_atom_type();
+              i_center++;
+            }
+            i_frame++;
+          }
+          return ij_mat;
+        },
+        R"(Get informations necessary to the computation of predictions. It has
+        as many rows as the number representations and they correspond to the
+        index of the structure, the central atom and its atomic species.
+        np.array with shape (n_atoms, 3))");
+
+    manager_collection.def(
+        "get_neighbors_for_gradient",
```

Comment:
This function name should be similar to the name of `get_atoms_for_predictions`. So maybe `get_gradients_metadata`, or something else if you have a better idea!

Reply:
renamed